File: src/leetCodeProblems/HashSearch/LongestConsecutiveSequence128.java
package leetCodeProblems.HashSearch;

/**
 * LeetCode - https://leetcode.com/problems/longest-consecutive-sequence/
 */

import java.util.*;

public class LongestConsecutiveSequence128 {
	public int longestConsecutive(int[] nums) {
        HashSet<Integer> hashSet = new HashSet<Integer>();
        int ans = 0;

        for(int i=0; i < nums.length; i++) {
            hashSet.add(nums[i]);
        }

        for(int i=0; i < nums.length; i++) {

            // This is important, to remove UNREQUIRED loops.
            if (!hashSet.contains(nums[i]-1)) {

                int j = 0;

                while(hashSet.contains(nums[i]+j)) {
                    j++;
                }

                if (j > ans) {
                    ans = j;
                }
            }
        }

        return ans;
    }
}

File: src/leetCodeProblems/HashMap/LongestConsecutiveSequence128.go
package main

/*
- Leetcode - https://leetcode.com/problems/longest-consecutive-sequence
- Time - O(n)
- Space - O(n)
*/
func longestConsecutive(nums []int) int {
	m := make(map[int]bool)

	for _, v := range nums {
		m[v] = true
	}

	out := 0

	for _, v := range nums {
		if ok := m[v-1]; !ok {
			j := 0

			for m[v+j] {
				j++
			}

			if j > out {
				out = j
			}

			if j > len(nums)/2 {
				break
			}
		}
	}
	return out
}
